sycl/doc/extensions/EnqueueBarrier/enqueue_barrier.asciidoc

```diff
@@ -0,0 +1,315 @@
+  // CG3
+});
+
+Queue.submit([&](cl::sycl::handler& cgh) {
+  cgh.barrier();
+});
+
+Queue.submit([&](cl::sycl::handler& cgh) {
+  // CG4
+});
+...
+----
+
+==== 2. Using `queue::submit_barrier()`:
+
+[source,c++,NoName,linenums]
+----
+...
+Queue.submit([&](cl::sycl::handler& cgh) {
+  // CG1
+});
+Queue.submit([&](cl::sycl::handler& cgh) {
+  // CG2
+});
+Queue.submit([&](cl::sycl::handler& cgh) {
+  // CG3
+});
+
+Queue.submit_barrier();
+
+Queue.submit([&](cl::sycl::handler& cgh) {
+  // CG4
+});
+...
+----
+
+
+=== Scenario 2: Enqueued barrier dependent on specific events from previously submitted commands
+
+CG3 requires CG1 (in Queue1) and CG2 (in Queue2) to have completed before it (CG3) begins execution.
+
+==== 1. Using `handler::barrier()`:
+
+[source,c++,NoName,linenums]
+----
+...
+auto event_barrier1 = Queue1.submit([&](cl::sycl::handler& cgh) {
+  // CG1
+});
+
+auto event_barrier2 = Queue2.submit([&](cl::sycl::handler& cgh) {
+  // CG2
+});
+
+Queue3.submit([&](cl::sycl::handler& cgh) {
+  cgh.barrier( vector_class<event>{event_barrier1, event_barrier2} );
+});
+
+Queue3.submit([&](cl::sycl::handler& cgh) {
+  // CG3
+});
+...
+----
+
+==== 2. Using `queue::submit_barrier()`:
+
+[source,c++,NoName,linenums]
+----
+...
+auto event_barrier1 = Queue1.submit([&](cl::sycl::handler& cgh) {
+  // CG1
+});
+
+auto event_barrier2 = Queue2.submit([&](cl::sycl::handler& cgh) {
+  // CG2
+});
+
+Queue3.submit_barrier( vector_class<event>{event_barrier1, event_barrier2} );
+
+Queue3.submit([&](cl::sycl::handler& cgh) {
+  // CG3
+});
+...
+----
+
+== Specification changes
+
+=== Modify part of Section 4.6.5.1
+
+*Change from:*
+[source,c++,NoName,linenums]
+----
+...
+template <typename T>
+event submit(T cgf, const queue &secondaryQueue);
+
+void wait();
+...
+----
+*To:*
+[source,c++,NoName,linenums]
+----
+...
+template <typename T>
+event submit(T cgf, const queue &secondaryQueue);
+
+event submit_barrier();
+
+event submit_barrier( const vector_class<event> &waitList );
+
+void wait();
+...
+----
+=== Add rows to Table 4.22
+
+[cols="70,300"]
+[grid="rows"]
+[options="header"]
+|========================================
+|*Member functions*|*Description*
+|`event submit_barrier()` | Same effect as submitting a `handler::barrier()` within a command group to this `queue`.  The returned event enters the `info::event_command_status::complete` state when all events that the barrier is dependent on (implicitly from all previously submitted commands to the same queue) have entered the `info::event_command_status::complete` state.
+|`event submit_barrier( const vector_class<event> &waitList )` | Same effect as submitting a `handler:barrier( const vector_class<event> &waitList )` within a command group to this `queue`.  The returned event enters the `info::event_command_status::complete` state when all events that the barrier is dependent on (explicitly from `waitList`) have entered the `info::event_command_status::complete` state.
```

Comment:
Can any of these methods throw any exception? e.g. should it throw an exception if the waitList is empty? is there any known condition where the barrier will fail to execute? If that is the case, is that a synchronous exception or asynchronous (via callback?)

Reply:
I don't have strong opinion. Any comments? @mkinsner

Reply:
I don't think there should be an exception on empty `waitList` because there are good use cases for a dynamic list that we want to enable, and the list can happen to be empty.

I suppose that there could be error conditions if an event in `waitList` is invalid, but SYCL generally doesn't define these sorts of error conditions.  So in short, I'm not aware of any condition that we should define which would throw an exception from this API, and therefore I'd choose to not say anything about them in this version of the extension.